Bhavya Singhal

B071

70022000199

BTech CE 

In [ ]:
import tensorflow as tf
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
df= pd.read_csv('/content/iris_data.csv')

In [ ]:
df.head

<bound method NDFrame.head of      sepal_length  sepal_width  petal_length  petal_width         species
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]>

In [ ]:
from sklearn.preprocessing import OneHotEncoder


In [ ]:
label = np.array(df.iloc[:,-1])   #all rows and (-1) -> last column
label.shape

(150,)

In [ ]:
label=label.reshape((-1,1))     #(-1) -> all rows but 1 column only
label.shape

(150, 1)

In [ ]:
one_hot = OneHotEncoder(sparse=False)

In [ ]:
y_onehot = one_hot.fit_transform(label)

In [ ]:
y_onehot.shape

(150, 3)

In [ ]:
dataset = pd.DataFrame({'flower1': y_onehot[:,0], 'flower2': y_onehot[:,1], 'flower3': y_onehot[:,2]})

In [ ]:
dataset.head()

flower1  flower2  flower3
0      1.0      0.0      0.0
1      1.0      0.0      0.0
2      1.0      0.0      0.0
3      1.0      0.0      0.0
4      1.0      0.0      0.0

In [ ]:
df.drop('species', axis=1, inplace=True)

In [ ]:
final =  pd.concat([df,dataset], axis =1)

In [ ]:
final

sepal_length  sepal_width  petal_length  petal_width  flower1  flower2  \
0             5.1          3.5           1.4          0.2      1.0      0.0   
1             4.9          3.0           1.4          0.2      1.0      0.0   
2             4.7          3.2           1.3          0.2      1.0      0.0   
3             4.6          3.1           1.5          0.2      1.0      0.0   
4             5.0          3.6           1.4          0.2      1.0      0.0   
..            ...          ...           ...          ...      ...      ...   
145           6.7          3.0           5.2          2.3      0.0      0.0   
146           6.3          2.5           5.0          1.9      0.0      0.0   
147           6.5          3.0           5.2          2.0      0.0      0.0   
148           6.2          3.4           5.4          2.3      0.0      0.0   
149           5.9          3.0           5.1          1.8      0.0      0.0   

     flower3  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
..       ...  
145      1.0  
146      1.0  
147      1.0  
148      1.0  
149      1.0  

[150 rows x 7 columns]

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [ ]:
std_scaler = StandardScaler()
final.iloc[:,[0,1,2,3]] = std_scaler.fit_transform(final.iloc[:,[0,1,2,3]])

In [ ]:
final.head()

sepal_length  sepal_width  petal_length  petal_width  flower1  flower2  \
0     -0.900681     1.032057     -1.341272    -1.312977      1.0      0.0   
1     -1.143017    -0.124958     -1.341272    -1.312977      1.0      0.0   
2     -1.385353     0.337848     -1.398138    -1.312977      1.0      0.0   
3     -1.506521     0.106445     -1.284407    -1.312977      1.0      0.0   
4     -1.021849     1.263460     -1.341272    -1.312977      1.0      0.0   

   flower3  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      0.0

In [ ]:
def sig(x):
  sig_x = 1/(1+np.exp(-x))
  return sig_x

In [ ]:
def accuracy(y_pred, y_true):
  acc = y_pred.argmax(axis=1)==y_true.argmax(axis=1)
  temp = acc.mean()
  return temp 

In [ ]:
def msef(y_pred, y_true):
  temp= (((y_pred-y_true)**2).sum())/(y_pred.size)
  return temp

In [ ]:
input = final.iloc[:,0:4]

In [ ]:
N = input.size
iterations =5000
learning_rate=0.01
input_size =4
hidden_size =2
output_size = 3
result1 = pd.DataFrame(columns= ['mse', 'accuracy'])

In [ ]:
np.random.seed(10)
w1 = np.random.normal(scale=0.5, size = (input_size, hidden_size))
w2 = np.random.normal(scale=0.5, size = (hidden_size, output_size))
bias1 = np.random.rand(1)
bias2 = np.random.rand(1)

In [ ]:
for i in range(iterations):
  z1 = np.dot(input,w1)+bias1
  a1 = sig(z1)
  z2 = np.dot(a1,w2)
  a2 = sig(z2)
  mse = msef(a2, y_onehot)
  acc = accuracy(a2, y_onehot)
  result1 = result1.append({'mse':mse, 'accuracy':acc}, ignore_index=True)
  e1 = a2 - y_onehot
  #partial derivative of output
  dw1=e1*a2*(1-a2)
  e2 = np.dot(dw1, w2.T)
  dw2=e2*a1*(1-a1)
  w2_update = np.dot(a1.T,dw1)/N
  w1_update = np.dot(input.T,dw2)/N
  w2 = w2 - learning_rate*w2_update
  w1 = w1 - learning_rate*w1_update

In [ ]:
result1.shape

(5000, 2)

In [ ]:
result1

mse  accuracy
0     0.274659  0.513333
1     0.274640  0.513333
2     0.274622  0.513333
3     0.274603  0.513333
4     0.274584  0.513333
...        ...       ...
4995  0.218315  0.560000
4996  0.218308  0.560000
4997  0.218302  0.560000
4998  0.218296  0.560000
4999  0.218290  0.560000

[5000 rows x 2 columns]

Conclusion:

3 layer network with 1 hidden layer is built for classification of iris dataset. The given network has only 2 neurons in hidden layer. For 100 iterations and LR=0.01 MSE = 0.272832, Accuracy = 0.500000. If No. of iterations is increased to 5000 then MSE = 0.218290 and accuracy = 0.560000. 

Since the network is too simple, the accuracy is very low. To improve the accuracy the number of hidden layers and number of hidden neurons in each layer must be increased.

If random state is changed to some other value, there is no appreciable change in error and accuracy. This is because the network can be initialised to any value, it is the backpropagation algorithm which changes the values to reach the optimum value.